net/udprelay/server.go

```diff
@@ -337,37 +338,49 @@ func NewServer(logf logger.Logf, port uint16, onlyStaticAddrPorts bool) (s *Serv
 		Logf:   logger.WithPrefix(logf, "netcheck: "),
 		SendPacket: func(b []byte, addrPort netip.AddrPort) (int, error) {
 			if addrPort.Addr().Is4() {
-				return s.uc4.WriteToUDPAddrPort(b, addrPort)
-			} else if s.uc6 != nil {
-				return s.uc6.WriteToUDPAddrPort(b, addrPort)
+				return s.uc4[0].WriteToUDPAddrPort(b, addrPort)
+			} else if len(s.uc6) > 0 {
+				return s.uc6[0].WriteToUDPAddrPort(b, addrPort)
 			} else {
 				return 0, errors.New("IPv6 socket is not bound")
 			}
 		},
 	}
 
-	err = s.listenOn(port)
+	err = s.bindSockets(port)
 	if err != nil {
 		return nil, err
 	}
+	s.startPacketReaders()
 
 	if !s.onlyStaticAddrPorts {
 		s.wg.Add(1)
 		go s.addrDiscoveryLoop()
 	}
 
-	s.wg.Add(1)
-	go s.packetReadLoop(s.uc4, s.uc6, true)
-	if s.uc6 != nil {
-		s.wg.Add(1)
-		go s.packetReadLoop(s.uc6, s.uc4, false)
-	}
 	s.wg.Add(1)
 	go s.endpointGCLoop()
 
 	return s, nil
 }
 
+func (s *Server) startPacketReaders() {
+	for _, readSockets := range []*[]batching.Conn{&s.uc4, &s.uc6} {
+		for i, uc := range *readSockets {
+			otherSockets := &s.uc6
+			if readSockets == &s.uc6 {
+				otherSockets = &s.uc4
+			}
+			var other batching.Conn
+			if len(*otherSockets) > 0 {
+				other = (*otherSockets)[min(len(*otherSockets)-1, i)]
+			}
+			s.wg.Add(1)
```

Comment:
Could we use `s.wg.Go` instead of maintaining the counter explicitly?

Reply:
No, unfortunately. See the event order in the `defer func()` in `Server.packetReadLoop`. Could probably be refactored in a separate PR to simplify.